### PR TITLE
Replicaof bug fix

### PR DESCRIPTION
**Pull request by OfirMos**
Fixing bug:
https://redislabs.atlassian.net/browse/RED-78216?focusedCommentId=1589712&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-1589712

In general keep cross reference on restore is wrong, there is not guarantee that the other key will ever be restored or that it might be renamed.
That said, on enterprise replica of we do want to keep the cross reference
The solution: 
Use persistence subevent to know when other than dump rdb_save is in progress and removing the references when the rdb_save is in dump context.
Inorder to handle old rdb versions on which the references exists on dump rdbs, we save the last rdb load version so that on restore keyspace notification we can tell if we should remove the cross references.

There is a limitation in the solution, which is when updating a key while replicaof is in progress the downsampled key might not be updated because it might not replicated yet.

### CODE REVIEW

File: ramp.yml [hunk 1]
```diff
@@ -22,3 +22,5 @@ capabilities:
     - intershard_tls
     - intershard_tls_pass
     - flash
+exclude_commands:
+    - ts._replicate
```
Comment (MeirShpilraien) [inline ramp.yml:26]: Shouldn't it be `ts._replicate_auxLoad_before` and `ts._replicate_auxLoad_after`?

File: src/rdb.c [hunk 2]
```diff
@@ -219,3 +222,51 @@ void series_rdb_save(RedisModuleIO *io, void *value) {
...
+
+int aux_load_count = 0;
+
+int auxLoad(RedisModuleIO *io, int encver, int when) {
+    if (REDISMODULE_AUX_BEFORE_RDB) {
+        replicaof_in_progress = true;
+        aux_load_count++;
+
+        // replicate auxLoad (fixes a bug in Redis < 6.2.10 auxLoad isn't replicated)
+        RedisModuleCallReply *reply =
```
Comment (MeirShpilraien) [inline src/rdb.c:234]: Why not just use RM_Replicate? We say that this is what we need to check, if it is OK to call it here...

File: tests/flow/test_ts_madd.py [hunk 3]
```diff
@@ -130,7 +140,13 @@ def test_madd_some_failed_replicas():
         env.assertEqual(r.execute_command("ts.range", "test_key1", "-", "+"),  [[122, b'11'], [123, b'11'], [124, b'12'], [125, b'12']])
 
     if env.useAof and isinstance(env.envRunner, StandardEnv):
-        cmds = aof_parser.parse_file(os.path.join(env.envRunner.dbDirPath, env.envRunner._getFileName('master', '.aof')))
+        dbDirPath = env.envRunner.dbDirPath
+        if not is_less_than_ver7:
```
Comment (MeirShpilraien) [inline tests/flow/test_ts_madd.py:144]: I am not sure I understand this differentiation to Redis 7, please explain and add a comment.

**Review by MeirShpilraien — COMMENTED** [review]
Looks good but I think we are missing some important details:
1. I do not see why you need to use RM_Call, I also believe that the code that checks RM_Call return type is wrong. We can use RM_Replicate directly.
2. We do not always need to replicate the aux callback, we need to check the Redis version to understand if it has the aux bug.
3. I do not see you register to flushall event to clear `aux_load_count`.
4. I do not see you block all commands when `aux_load_count > 0`.
5. I thing `replicaof_in_progress` is redudent.
6. I do not fully understand the test, I think we should cover it better.

**Comment by OfirMos** [comment]
@MeirShpilraien 
"4. I do not see you block all commands when aux_load_count > 0."
This is a breaking change, @LiorKogan was against adding this.

File: src/module.c [hunk 4]
```diff
@@ -1120,6 +1120,26 @@ void swapDbEventCallback(RedisModuleCtx *ctx, RedisModuleEvent e, uint64_t sub,
     }
 }
 
+int persistence_in_progress = 0;
+
+void persistCallback(RedisModuleCtx *ctx, RedisModuleEvent eid, uint64_t subevent, void *data) {
+    if (memcmp(&eid, &RedisModuleEvent_Persistence, sizeof(eid)) != 0) {
```
Comment (MeirShpilraien) [inline src/module.c:1126]: I do not think this is needed, you register this callback on the Persistence event only.

**Review by MeirShpilraien — COMMENTED** [review]
LGTM, please make sure to update the top comment with the solution limitations (the fact that if a key is updated during the rdb phase of the replica of, it will failed).
I would like @danni-m to also review this.

File: src/module.c (hunk 4 above)
Comment (OfirMos) [reply src/module.c:1126]: I think it's more safe to make sure it's the type of event, don't you think?

File: src/module.c [hunk 5]
```diff
@@ -1120,6 +1120,26 @@ void swapDbEventCallback(RedisModuleCtx *ctx, RedisModuleEvent e, uint64_t sub,
     }
 }
 
+int persistence_in_progress = 0;
```
Comment (danni-m) [inline src/module.c:1123]: @OfirMos do we use an integer because we are expecting to have more than 1 persistence flow at the same time?

**Review by danni-m — APPROVED** [review]

File: src/module.c (hunk 5 above)
Comment (OfirMos) [reply src/module.c:1123]: @danni-m Yes, I'm not expecting it to grow more than 1 but it's for safeties

**Pull request merged**